### DISCUSSION TITLE
Telnet/ssh support

### DISCUSSION THREAD

**Discussion by bvarg** [tui-cs/Terminal.Gui#1601] [Ideas]
I need to run the console application rendered as telnet app. Is this  doable using this library? If yes is there any samples or anyone can help me directing on how to achieve this.

**Comment by tznind** [comment]
Do you need to use telnet specifically and not just SSH? SSH works out of the box

Implementing telnet I think would be a lot of work.  Unless I am misunderstanding you would be wanting to have a TCPListener and send the strings yourself kind of like this https://stackoverflow.com/questions/19169893/simple-telnet-console-that-listens-and-accepts-connections

It might be possible by implementing your own [ConsoleDriver](https://github.com/migueldeicaza/gui.cs/blob/main/Terminal.Gui/Core/ConsoleDriver.cs#L631).  I would start by looking at the unit testing driver [FakeConsole](https://github.com/migueldeicaza/gui.cs/blob/main/Terminal.Gui/ConsoleDrivers/FakeDriver/FakeConsole.cs) if you want to go down that route.

**Reply by JessicaMulein** [reply]
I was writing one for Spectre.Console to do a web based console... I'm thinking I may resume that effort, but on Gui.cs...

**Reply by bvarg** [reply]
I need to use telnet specifically. I should be be able to run the console screen in a telnet session and yes right I definitely need to have a TCPListner. But I am not sure how do I transform the controls to render GUI controls on Telnet screen? For example I need to design a data entry screen on Telnet console to accept user input

**Reply by tznind** [reply]
> But I am not sure how do I transform the controls to render GUI controls on Telnet screen? For example I need to design a data entry screen on Telnet console to accept user input

I guess there are 3 ways you could go with this:
- If there is a library or application that presents a curses compatible console over telnet then you should just be able to run the Terminal.gui app directly
- You could write your application normally and then start the process from a wrapper C# application which rebinds standard in and standard out to itself and translates that into whatever telnet sessions are open
- You could create all the Terminal.Gui controls as normal and a custom `ConsoleDriver` that has a screen buffer and handles translating state into whatever open telnet connection is being serviced.

If your plan is to do the third of those then I would look at the tests which use `FakeDriver`.  It is used in tests to check the console screen characters after redrawing without any blocking or threading.  Here is a good example of this:

https://github.com/migueldeicaza/gui.cs/blob/a822e1afa948531a6879c2f5888c29ec863d81da/UnitTests/GraphViewTests.cs#L1201-L1242

Since these classes are internal I guess you would have start by cloning this repo and testing out how feasible it is in the main project or as a new project with a reference to the gui.cs test project.

But it sounds like @FreddieMercurial  has a better idea of what is involved in this, I may be not understanding exactly your requirements.

**Comment by JessicaMulein** [comment]
@tznind seems right on.

In my own code quest, I've seen a C# telnet/ssh terminal code example.
My personal quest is to implement a Web based HTML5 console so I can do web games and apps.
I have been working on other projects for a while, but this has re-awakened my interest.

This and the other nearby partials.
https://github.com/HACC-ers/HACC/blob/main/Models/Drivers/Html5AnsiConsoleCanvas.TerminalGuiConsoleDriver.cs

It's starting to look like actually a fair chunk of my old code is going to be rebuilt to work around the TerminalGui core instead of the Spectre console core- it seems like it is a wiser implementation to make the Spectre just use the TerminalGui core in my case. which is not at all relevant to you- but maybe something in my code will help?

PS this was the project I recalled
https://www.codeproject.com/Articles/19071/Quick-tool-A-minimalistic-Telnet-library
Notably it has no cursor/graphics control, but it may be part of the missing pieces you need.

**Comment by JessicaMulein** [comment]
Actually, @bvarg maybe you should check out the NetDriver? https://github.com/migueldeicaza/gui.cs/blob/main/Terminal.Gui/ConsoleDrivers/NetDriver.cs

**Reply by bvarg** [reply]
Will check this out and see

**Comment by BDisp** [comment]
@FreddieMercurial if the use of `Terminal.Gui` is useful, as I think it is, then the most easy way to implement a custom console, like `WebConsole` or `TelnetConsole`, in the case where the `NetDriver` don't satisfies your needs, is by creating a new `ConsoleDriver` derived class. In this case the place to initialize one of those console is by creating bool variables on `Application` class to load the requested console.
As the `UseSystemConsole` variable is used to load the `NetDriver`, a `UseWebConsole` and a `UseTelnetConsole` will be used to load the `WebDriver` and `TelnetDriver`, respectively. This variables must be checked first and if all of them are false, then the `WindowsDriver` will be loaded for `Windows` environment and the `CursesDriver` for the `Linux` and `macOS`. See this:

https://github.com/migueldeicaza/gui.cs/blob/a822e1afa948531a6879c2f5888c29ec863d81da/Terminal.Gui/Core/Application.cs#L339-L354


@FreddieMercurial in this case you can use the internal `Clipboard` from the `Terminal.Gui`, without added more new libraries.
If you want reopen the issue #1638 and work together on a possible solution, I'm available.

**Reply by BDisp** [reply]
> It also looks like users of my WebConsole will need to call Application.Init with the correct WebConsoleDriver and the WebConsoleMainLoop

Yes, the WebConsoleDriver will be responsible for loading the right web libraries and the WebConsoleMainLoopwith deals with the call to the main iterations.

**Reply by BDisp** [reply]
Well, I'm curious now and I like to help but I need you to guide me on this. Then I suggest you fork the Gui and implementing the needed tools and I'll help you on what I've knowledge.
Maybe is better open a new issue with the feature you need and continue discussion there.

**Reply by JessicaMulein** [reply]
Good to their word, BDisp has helped move us along immensely.
I've pruned most of my comments.

**Comment by JessicaMulein** [comment]
@bvarg sorry for derailing your issue - we can clean it up maybe as we've moved the conversation elsewhere.
We've learned a lot about the Blazor application lifecycle, but I'm not sure how much we've learned relevant to your original question.
Feel free to look at the https://github.com/HACC-ers/HACC repo under the WebConsoleDriver partials.

**Reply by JessicaMulein** [reply]
PS if you have any questions- drop a note and @BDisp or I maybe can actually help more now :)

**Comment by migueldeicaza** [comment]
The original issue is quite ambiguous, Telnet and ssh are transports to connect to a terminal, and Gui.cs will work with any transport, as long as the endpoint can support some basic terminal commands, and the TERM environment variable is correctly configured.

**Reply by bvarg** [reply]
Just an update. I was able to run my console app through SSH.

**Reply by wzuqui** [reply]
I ended up finding this thread, but it had already created an issue #2723.

Any news about a driver for TELNET or something?
